internal/datastore/crdb/caveat.go

```diff
@@ -2,24 +2,189 @@ package crdb
+		colCaveatDefinition,
+	)
+	writeCaveat  = psql.Insert(tableCaveat).Columns(colCaveatName, colCaveatDefinition).Suffix(upsertCaveatSuffix)
+	readCaveat   = psql.Select(colCaveatDefinition, colTimestamp).From(tableCaveat)
+	listCaveat   = psql.Select(colCaveatName, colCaveatDefinition).From(tableCaveat).OrderBy(colCaveatName)
+	deleteCaveat = psql.Delete(tableCaveat)
+)
+
+const (
+	errWriteCaveat   = "unable to write new caveat revision: %w"
+	errReadCaveat    = "unable to read new caveat `%s`: %w"
+	errListCaveats   = "unable to list caveat: %w"
+	errDeleteCaveats = "unable to delete caveats: %w"
+)
+
 func (cr *crdbReader) ReadCaveatByName(ctx context.Context, name string) (*core.CaveatDefinition, datastore.Revision, error) {
-	return nil, datastore.NoRevision, fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(ctx, "ReadCaveatByName", trace.WithAttributes(attribute.String("name", name)))
+	defer span.End()
+
+	query := readCaveat.Where(sq.Eq{colCaveatName: name})
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errReadCaveat, name, err)
+	}
+
+	var definitionBytes []byte
+	var timestamp time.Time
+	err = cr.executeWithTx(ctx, func(ctx context.Context, tx pgx.Tx) error {
+		if err := tx.QueryRow(ctx, sql, args...).Scan(&definitionBytes, &timestamp); err != nil {
+			if errors.Is(err, pgx.ErrNoRows) {
+				err = datastore.NewCaveatNameNotFoundErr(name)
+			}
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errReadCaveat, name, err)
+	}
+	loaded := &core.CaveatDefinition{}
+	if err := loaded.UnmarshalVT(definitionBytes); err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errReadCaveat, name, err)
+	}
+	cr.addOverlapKey(name)
+	return loaded, revisionFromTimestamp(timestamp), nil
 }
 
 func (cr *crdbReader) ListCaveats(ctx context.Context, caveatNames ...string) ([]*core.CaveatDefinition, error) {
-	return nil, fmt.Errorf("unimplemented caveat support in datastore")
+	ctx, span := tracer.Start(ctx, "ListCaveats", trace.WithAttributes(attribute.StringSlice("names", caveatNames)))
+	defer span.End()
+
+	caveatsWithNames := listCaveat
+	if len(caveatNames) > 0 {
+		caveatsWithNames = caveatsWithNames.Where(sq.Eq{colCaveatName: caveatNames})
+	}
+
+	sql, args, err := caveatsWithNames.ToSql()
+	if err != nil {
+		return nil, fmt.Errorf(errListCaveats, err)
+	}
+	var allDefinitionBytes [][]byte
+	var overlappingKeys []string
+	err = cr.executeWithTx(ctx, func(ctx context.Context, tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, sql, args...)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			var defBytes []byte
+			var overlappingKey string
+			err = rows.Scan(&overlappingKey, &defBytes)
+			if err != nil {
+				return err
+			}
+			allDefinitionBytes = append(allDefinitionBytes, defBytes)
+			overlappingKeys = append(overlappingKeys, overlappingKey)
```

Comment:
```suggestion
			cr.addOverlapKey(val)
```
this bit is still double-tracking the keys

Reply:
@ecordell whoops! Fixed!